test/framework/klusterlet.go

```diff
@@ -0,0 +1,259 @@
+				{
+					URL: "https://localhost",
+				},
+			},
+			ClusterName: clusterName,
+			Namespace:   klusterletNamespace,
+			DeployOption: operatorapiv1.KlusterletDeployOption{
+				Mode: mode,
+			},
+		},
+	}
+
+	agentNamespace := helpers.AgentNamespace(klusterlet)
+	klog.Infof("klusterlet: %s/%s, \t mode: %v, \t agent namespace: %s", klusterlet.Name, klusterlet.Namespace, mode, agentNamespace)
+
+	// create agentNamespace
+	namespace := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: agentNamespace,
+			Annotations: map[string]string{
+				"workload.openshift.io/allowed": "management",
+			},
+		},
+	}
+	if _, err := spoke.KubeClient.CoreV1().Namespaces().Get(context.TODO(), agentNamespace, metav1.GetOptions{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.Errorf("failed to get ns %v. %v", agentNamespace, err)
+			return nil, err
+		}
+
+		if _, err := spoke.KubeClient.CoreV1().Namespaces().Create(context.TODO(),
+			namespace, metav1.CreateOptions{}); err != nil {
+			klog.Errorf("failed to create ns %v. %v", namespace, err)
+			return nil, err
+		}
+	}
+
+	// create bootstrap-hub-kubeconfig secret
+	secret := bootstrapHubKubeConfigSecret.DeepCopy()
+	if _, err := spoke.KubeClient.CoreV1().Secrets(agentNamespace).Get(context.TODO(), secret.Name, metav1.GetOptions{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.Errorf("failed to get secret %v in ns %v. %v", secret.Name, agentNamespace, err)
+			return nil, err
+		}
+		if _, err = spoke.KubeClient.CoreV1().Secrets(agentNamespace).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
+			klog.Errorf("failed to create secret %v in ns %v. %v", secret, agentNamespace, err)
+			return nil, err
+		}
+	}
+
+	if helpers.IsHosted(mode) {
+		// create external-managed-kubeconfig, will use the same cluster to simulate the Hosted mode.
+		secret.Namespace = agentNamespace
+		secret.Name = helpers.ExternalManagedKubeConfig
+		if _, err := spoke.KubeClient.CoreV1().Secrets(agentNamespace).Get(context.TODO(), secret.Name, metav1.GetOptions{}); err != nil {
+			if !apierrors.IsNotFound(err) {
+				klog.Errorf("failed to get secret %v in ns %v. %v", secret.Name, agentNamespace, err)
+				return nil, err
+			}
+			if _, err = spoke.KubeClient.CoreV1().Secrets(agentNamespace).Create(context.TODO(), secret, metav1.CreateOptions{}); err != nil {
+				klog.Errorf("failed to create secret %v in ns %v. %v", secret, agentNamespace, err)
+				return nil, err
+			}
+		}
+	}
+
+	// create klusterlet CR
+	realKlusterlet, err := spoke.OperatorClient.OperatorV1().Klusterlets().Create(context.TODO(),
+		klusterlet, metav1.CreateOptions{})
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		klog.Errorf("failed to create klusterlet %v . %v", klusterlet.Name, err)
+		return nil, err
+	}
+
+	return realKlusterlet, nil
+}
+
+func CreatePureHostedKlusterlet(spoke *Spoke, name, clusterName string) (*operatorapiv1.Klusterlet, error) {
```

Comment:
why this func, and why this is different from CreateKlusterlet. When should we use this?

Reply:
That's a good question but may beyond this PR's scope, the function is directly copied from Tester and I will try to answer.

Comparing with the `CreateKlusterlet` which not only create Klusterlet but also create the related reasources like namespace, bootstrapkubeconfigsecret, externalKubeconfigSecret, the `CreatePureHostedKlusterlet ` only create Klusterlet, no another thing.

It seems there is a case that requires only Hosted mode Klusterlet, but don't want the external managed kubeconfig, that's why we want a 'Pure' function.
```
	It("Delete klusterlet CR in Hosted mode without external managed kubeconfig", func() {
		By(fmt.Sprintf("create klusterlet %v with managed cluster name %v in Hosted mode", klusterletName, clusterName))
		_, err := CreatePureHostedKlusterlet(spoke, klusterletName, clusterName)
		Expect(err).ToNot(HaveOccurred())
```

I can add a TODO on this function and the `CreateKlusterlet` to try to merge them into one if we need.